File: C/check2.c

#include <stdio.h>
#include <string.h>

#define FOR(i,a,b) for(int i=a;i<b;i++)
#define REP(i,n)  FOR(i,0,n)

#define MAXCITY 100
#define MAXCOUNT 2000 + 1
char city_name[MAXCITY][18]; //駅名の表
int ncity; //駅数
int nconn; //列車数

// 列車配列
struct train {
	int from, to;
	int dpt, arv;
	int fare;
} trains[MAXCOUNT];

//駅名 -> 駅番号
int city_id(char *name) {
	FOR(i,0,ncity) {
		if(strcmp(name, &(city_name[i][0])) == 0) return i;
	}
	strcpy(&(city_name[ncity][0]), name);
	return ncity++;
}

// 列車追加
void parse_connection(char *buf) {
	char from[18], to[18];
	int dpt[2], arv[2], fare;

	sscanf(buf,"%s %d:%d %s %d:%d %d",
			from, &(dpt[0]), &(dpt[1]), to, &(arv[0]), &(arv[1]), &fare);

	trains[nconn].from = city_id(from);
	trains[nconn].to = city_id(to);
	trains[nconn].dpt = dpt[0]*60 + dpt[1];
	trains[nconn].arv = arv[0]*60 + arv[1];
	trains[nconn].fare = fare;
	nconn++;
}

int main() {
	int count, maxFare, maxTime, time;
	char buf[64];

	while(fgets(buf, sizeof(buf), stdin) != NULL) {
		// パラメータの初期化
		nconn = 0;
		ncity = 0;
		maxFare = 0;
		maxTime = 0;

		// 入力行数取得
		sscanf(buf, "%d", &count);
		printf("%d\n", count);

		// 0行なら終了
		if(!count) break;

		// 列車の追加
		REP(i,count) {
			if(fgets(buf, sizeof(buf), stdin) == NULL) break;
			parse_connection(buf);
		}

		//最大値チェック
		REP(i,count) {
			if(maxFare < trains[i].fare) maxFare = trains[i].fare;
			time = trains[i].arv - trains[i].dpt;
			if(maxTime < time) maxTime = time;
		}

		// 最大値の列車出力
		printf("Max fare: %d yen Max fare id: ",maxFare);
		REP(i,count) if(maxFare == trains[i].fare) printf("%d ",i);
		printf("\n");
		printf("Max time: %d min Max time id: ",maxTime);
		REP(i,count) if(maxTime == trains[i].arv - trains[i].dpt) printf("%d ",i);
		printf("\n");

		// 盛岡行きの列車を出力
		REP(i,count) {
			if(trains[i].to == city_id("Morioka"))
				printf("%d %d %d %d %d\n",
					trains[i].from,trains[i].dpt,trains[i].to,trains[i].arv,trains[i].fare);
		}
	}
	return 0;
}

File: Python/check2.py

import sys
sys.stdin = open('../input.txt')

# MAXCITY = 100
# MAXCOUNT = 2000 + 1
city_name = {}
trains = []
ncity = 0
nconn = 0

# 駅名 -> 駅番号
def city_id(name):
    global ncity, nconn

    # keyが存在すればvalueを返す
    if name in city_name:
        return city_name[name]

    # 要素を追加
    city_name[name] = ncity
    ncity += 1
    return ncity - 1

# 列車追加
def parse_connection(buf):
    global ncity, nconn

    # リスト内に辞書を追加
    trains.append({})
    trains[nconn]["from"] = city_id(buf[0])
    h,m = map(int, buf[1].split(":"))
    trains[nconn]["dpt"] = h*60 + m
    trains[nconn]["to"] = city_id(buf[2])
    h,m = map(int, buf[3].split(":"))
    trains[nconn]["arv"] = h*60 + m
    trains[nconn]["fare"] = int(buf[4])
    nconn += 1

if __name__ == '__main__':
    while True:
        ncity = 0
        nconn = 0
        trains.clear()
        city_name.clear()
        maxFare = 0
        maxTime = 0

        # 列車数読み込み
        count = int(input())
        print(count)

        # 終了条件
        if count == 0: break

        # 列車読み込み
        for i in range(0,count):
            buf = input().split()
            parse_connection(buf)

        # 最大値チェック
        for i in range(0,nconn):
            if maxFare < trains[i]["fare"]:
                maxFare = trains[i]["fare"]
            time = trains[i]["arv"] - trains[i]["dpt"]
            if maxTime < time:
                maxTime = time

        # 最大値及びその列車番号出力
        print("Max fare: ",maxFare," yen Max fare id: ",end = "")
        for i in range(0,nconn):
            if trains[i]["fare"] == maxFare:
                print(i," ",end = "")
        print()
        print("Max time: ",maxTime," yen Max time id: ",end = "")
        for i in range(0,nconn):
            if trains[i]["arv"] - trains[i]["dpt"] == maxTime:
                print(i," ",end = "")
        print()

        # 列車情報出力
        for i in range(0,nconn):
            for j in trains[i]:
                print(trains[i][j]," ",end = "")
            print()

        print()
